support PackedSequence type for apply_for_tensors

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76265

support PackedSequence type for apply_for_tensors, some rnn modules outputs are PackedSequence types

Differential Revision: [D35862156](https://our.internmc.facebook.com/intern/diff/D35862156/)

zhaojuanmao committed:
support PackedSequence type for apply_for_tensors

support PackedSequence type for apply_for_tensors, some rnn modules outputs are PackedSequence types

Differential Revision: [D35862156](https://our.internmc.facebook.com/intern/diff/D35862156/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/fsdp/utils.py:
```diff
@@ -8,18 +9,21 @@
     """Recursively apply to all tensor in different kinds of container types."""
 
-    def apply(x: Union[torch.Tensor, Dict, List, Tuple, Set, OrderedDict]) -> Any:
+    def apply(x: Union[torch.Tensor, Dict, List, Tuple, Set, OrderedDict, PackedSequence]) -> Any:
```
curious, do we need to support custom types (e.g., by allowing users to pass in a function or by pickling `x` if it is none of the listed types here?)?

zhaojuanmao replied:
the apply_to_tensors is mainly used for casting input tensors to mixed precision or register hooks of output tensors internally inside FSDP. possibly it is hard to expose to users for passing a function or so. We also do not want to pickling X for non-tensor data, e..g, we do not want to cast scalars or register hooks of scalars. 

but I'm curious how PyTorch generally get tensors from a custom defined type...? if users' input or output is a custom defined type and it contains tensors, we may not "apply_to_tensors" as expected for this custom defined type right now

mrshenli replied:
> but I'm curious how PyTorch generally get tensors from a custom defined type...? if users' input or output is a custom defined type and it contains tensors, we may not "apply_to_tensors" as expected for this custom defined type right now

FWIW, there is no good solution today. What CUDA Future does is pickling the Python obj and extract all tensors during the pickling process, but that can be pretty slow.

mrshenli commented on test/distributed/fsdp/test_utils.py:
```diff
@@ -92,6 +93,23 @@ def test_replace_by_prefix(self):
+        x, h = rnn(x)
+        x = _apply_to_tensors(fill_fn, x)
+        x, _ = nn.utils.rnn.pad_packed_sequence(x)
+        assert torch.sum(x) == 0
```
any reason we are using `assert` instead of `self.assertEqual`?

rohan-varma reviewed (approved):
LG

zhaojuanmao committed:
Update on "support PackedSequence type for apply_for_tensors"


support PackedSequence type for apply_for_tensors, some rnn modules outputs are PackedSequence types

Differential Revision: [D35862156](https://our.internmc.facebook.com/intern/diff/D35862156/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "support PackedSequence type for apply_for_tensors"


support PackedSequence type for apply_for_tensors, some rnn modules outputs are PackedSequence types

Differential Revision: [D35862156](https://our.internmc.facebook.com/intern/diff/D35862156/)

[ghstack-poisoned]

zhaojuanmao commented:
@pytorchbot merge this please
